xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -373,33 +408,58 @@ private Set<String> getResourceKeys(XdsResourceType<?> xdsResourceType) {
     return resourceSubscribers.get(xdsResourceType).keySet();
   }
 
-  private void cleanUpResourceTimers() {
+  // cpcForThisStream is null when doing shutdown
+  private void cleanUpResourceTimers(ControlPlaneClient cpcForThisStream) {
+    List<String> authoritiesForCpc = getActiveAuthorities(cpcForThisStream);
+
     for (Map<String, ResourceSubscriber<?>> subscriberMap : resourceSubscribers.values()) {
       for (ResourceSubscriber<?> subscriber : subscriberMap.values()) {
-        subscriber.stopTimer();
+        if (cpcForThisStream == null || authoritiesForCpc.contains(subscriber.authority)) {
+          subscriber.stopTimer();
+        }
+      }
+    }
+  }
+
+  public ControlPlaneClient getOrCreateControlPlaneClient(ImmutableList<ServerInfo> serverInfos) {
+    if (serverInfos == null || serverInfos.isEmpty()) {
+      return null;
+    }
+
+    for (ServerInfo serverInfo : serverInfos) {
+      ControlPlaneClient cpc = getOrCreateControlPlaneClient(serverInfo);
+      if (cpc.isInBackoff()) {
+        continue;
       }
+      return cpc;
     }
+
+    // Everything existed and is in backoff so return null
+    return null;
```

Comment:
Returning `null` here is confusing. The name would not suggest it could return `null` ("you said you'd create it"). And `getOrCreateControlPlaneClient(ServerInfo)` doesn't return `null`. So it is very hard to read code using this correctly. For example, the `ControlPlaneClient cpc = getOrCreateControlPlaneClient(serverInfo);` in this function (properly) doesn't check the `null` return, but it is type-sensitive whether that is appropriate. I suggest renaming this method (potentially also combining it with `fallBackToCpc()` that I mention elsewhere).